internal/pkg/agent/application/upgrade/details/details.go

```diff
@@ -68,6 +68,7 @@ func NewDetails(targetVersion string, initialState State, actionID string) *Deta
 
 // SetState is a convenience method to set the state of the upgrade and
 // notify all observers.
+// Do NOT call SetState with StateFailed; call the Fail method instead.
```

Comment:
Why? So I'm guessing the Fail method is much more than just sugar syntax, am I right?

Reply:
That's right. The [`Fail()` method](https://github.com/elastic/elastic-agent/blob/72653ac2b0b57deedd276c3abc8d19f9085b3154/internal/pkg/agent/application/upgrade/details/details.go#L99-L117) sets some other fields besides `state` as well.  It also takes an `error` as an input parameter, otherwise I would've simply called it directly from within `SetState(s State)` if `s == StateFailed`. 

With this godoc line at least there is some warning to the user not to call `SetState(StateFailed)` but call `Fail(err)` instead.  I'm open to better ideas, though.